Fix #852: Allow scoping operation to a single activation ID

jandusil opened the pull request:
- Add a new column activation_id into table pa_operation
- Update OperationEntity
- Update OperationCreateRequest
- Update OperationServiceBehavior::createOperation
- Update OperationServiceBehavior::attemptApproveOperation
- Add tests
 - Update DDL, Liquibase, migration guide and REST API documentation

banterCZ commented on powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java:
```diff
@@ -791,6 +803,12 @@ private static boolean proximityCheckPassed(final ProximityCheckResult proximity
     }
 
+    private static boolean activationIdMatches(final OperationApproveRequest operationApproveRequest, String activationId) {
+        return activationId == null ||
+                (operationApproveRequest.getAdditionalData().get("activationId") != null &&
+                        operationApproveRequest.getAdditionalData().get("activationId").equals(activationId));
```
🤔 Couldn't it be simplified?

```suggestion
        return activationId == null || activationId.equals(operationApproveRequest.getAdditionalData().get("activationId"));
```

banterCZ commented on powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/OperationCreateRequest.java:
```diff
@@ -38,5 +38,6 @@ public class OperationCreateRequest {
     private String externalId;
     private final Map<String, String> parameters = new LinkedHashMap<>();
     private Boolean proximityCheckEnabled;
+    private String activationId;
```
1) May we distinguish for swagger what is optional and what is required, please?
2) Where it is validated?

romanstrobl commented on docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231112-add-activation-id.xml:
```diff
@@ -0,0 +1,17 @@
+        </preConditions>
+        <comment>Add activation_id column to pa_operation</comment>
+        <addColumn tableName="pa_operation">
+            <column name="activation_id" type="varchar(37)" />
```
It probably should, look at the rest of the schema.

romanstrobl reviewed (dismissed):
Let's fix the foreign key, other than this I have nothing to add.

jandusil commented:
Thank you for your comments. Few things that I have fixed extra while changing the documentation. 

In `WebServices-Methods.md` the `OperationCreateRequest` description was wrong. Missing `activationFlag` and the there was `applicationId` property instead of `applications`.

Moreover, I have added the Swagger annotation and added validation for `activationId`.

banterCZ commented on docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231112-add-activation-id.xml:
```diff
@@ -0,0 +1,37 @@
+        </preConditions>
+        <comment>Add activation_id column to pa_operation</comment>
+        <addColumn tableName="pa_operation">
+            <column name="activation_id" type="varchar(37)"/>
```
What was the reason not to introduce the constraint in a single changeset? I mean something like
```xml 
<column ...>
   <constraints ... />
</column>
```

jandusil replied:
I though It would be better to divide, but since it is a new column I agree.

romanstrobl commented on docs/PowerAuth-Server-1.6.0.md:
```diff
@@ -21,3 +21,13 @@ and `pa_application` or `pa_operation`. Make sure that `pa_operation_application
+
+Add a new column `activation_id` to the `pa_operation` table. This column is a foreign key that references
+the `activation_id` column in the `pa_activation` table. Storing the `activation_id` in the `pa_operation` table
+provides several enhancements for developers:
```
I would remove the `for developers` words, because this change also impacts users.

romanstrobl commented on powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehaviorTest.java:
```diff
@@ -0,0 +1,212 @@
+     * Verifies that the operation is correctly created and stored without an activation ID.
+     */
+    @Test
+    void testCreateOperationWithout() throws GenericServiceException {
```
`OperationWithout()`...? A word is missing probably.

romanstrobl commented on powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehaviorTest.java:
```diff
@@ -0,0 +1,212 @@
+     * Verifies that the operation is successfully approved even without an activation ID.
+     */
+    @Test
+    void testApproveOperationEntityWithoutActivationIdIdSuccess() throws GenericServiceException {
```
Typo: `IdId`.

romanstrobl reviewed (dismissed):
I only found several typos now, otherwise good job!

banterCZ commented on powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehaviorTest.java:
```diff
@@ -0,0 +1,212 @@
+
+        final OperationDetailResponse operationDetailResponse = tested.createOperation(request);
+        assertTrue(operationRepository.findOperation(operationDetailResponse.getId()).isPresent());
+        final OperationEntity savedEntity = operationRepository.findOperation(operationDetailResponse.getId()).get();
```
Do we have to call `findOperation` twice?

banterCZ reviewed (commented):
`arch_db_structure.png` should be regenerated.

banterCZ reviewed (approved):
Looks good.
